File: src/main/java/com/jakobmenke/bootrestgenerator/Templates.java
package com.jakobmenke.bootrestgenerator;

import java.io.InputStream;
import java.util.Objects;
import java.util.Scanner;

public class Templates {
    public String getResourceTemplate(String mainPackage, String entityName) {
        String mainPackageName = mainPackage.replaceAll("/", ".");
        String fileTemplate = getFile("templates/rest.resource.tmpl");
        fileTemplate = fileTemplate.replace("{{mainPackageName}}", mainPackageName);
        fileTemplate = fileTemplate.replace("{{entityName}}", entityName);
        fileTemplate = fileTemplate.replace("{{serviceName}}", entityName.toLowerCase());
        fileTemplate = fileTemplate.replace("{{restServicePrefix}}", "GlobalConstants.CONTEXT_PATH");

        return fileTemplate;
    }

    public String getDaoTemplate(String mainPackage, String entityName) {
        String mainPackageName = mainPackage.replaceAll("/", ".");
        String fileTemplate = getFile("templates/dao.tmpl");
        fileTemplate = fileTemplate.replace("{{mainPackageName}}", mainPackageName);
        fileTemplate = fileTemplate.replace("{{entityName}}", entityName);
        String camelRepoName = entityName.substring(0, 1).toLowerCase() + entityName.substring(1) + "Repository";
        fileTemplate = fileTemplate.replace("{{camelRepositoryName}}", camelRepoName);

        return fileTemplate;
    }

    public String getConstantsTemplate(String mainPackage, String entityName) {
        String mainPackageName = mainPackage.replaceAll("/", ".");
        String fileTemplate = getFile("templates/constants.tmpl");
        fileTemplate = fileTemplate.replace("{{mainPackageName}}", mainPackageName);

        return fileTemplate;
    }

    public String getRepositoryTemplate(String mainPackage, String entityName) {
        String mainPackageName = mainPackage.replaceAll("/", ".");
        String fileTemplate = getFile("templates/repository.tmpl");
        fileTemplate = fileTemplate.replace("{{mainPackageName}}", mainPackageName);
        fileTemplate = fileTemplate.replace("{{entityName}}", entityName);

        return fileTemplate;
    }

    private String getFile(String fileName) {
        StringBuilder result = new StringBuilder();
        ClassLoader classLoader = Templates.class.getClassLoader();
        InputStream in = classLoader.getResourceAsStream(fileName);

        try (Scanner scanner = new Scanner(Objects.requireNonNull(in))) {
            while (scanner.hasNextLine()) {
                String line = scanner.nextLine();
                result.append(line).append("\n");
            }
        }

        return result.toString();
    }

    public String getEntityTemplate(Entity entity, String mainPackage) {
        String mainPackageName = mainPackage.replaceAll("/", ".");
        String fileTemplate = getFile("templates/entity.tmpl");
        fileTemplate = fileTemplate.replace("{{mainPackageName}}", mainPackageName);
        fileTemplate = fileTemplate.replace("{{entityName}}", entity.getEntityName());
        fileTemplate = fileTemplate.replace("{{tableName}}", entity.getTableName());

        StringBuilder stringBuilder = new StringBuilder();
        for (ColumnToField column : entity.getColumns()) {
            String indentation = "    ";
            if (column.getDatabaseIdType() != null) {
                stringBuilder.append(indentation).append(column.getDatabaseIdType()).append("\n");
            }
            if (column.getDatabaseIdType() != null && column.getDatabaseIdType().equalsIgnoreCase("@ManyToOne")) {
                stringBuilder.append(indentation).append("@JoinColumn(name = \"").append(column.getDatabaseColumnName()).append("\")\n");
            } else {
                stringBuilder.append(indentation).append("@Column(name = \"").append(column.getDatabaseColumnName()).append("\")\n");
            }
            String accessModifier = "private";
            stringBuilder.append(indentation).append(accessModifier).append(" ").append(column.getJavaType()).append(" ").append(column.getCamelCaseFieldName()).append(";\n\n");
        }

        stringBuilder.append("}");

        return fileTemplate + stringBuilder.toString();
    }
}

File: src/main/kotlin/com/jakobmenke/bootrestgenerator/templates/Templates.kt
package com.jakobmenke.bootrestgenerator.templates

import com.jakobmenke.bootrestgenerator.dto.Entity
import com.jakobmenke.bootrestgenerator.utils.EntityToRESTConstants
import com.jakobmenke.bootrestgenerator.utils.Globals
import java.util.Scanner

class Templates {
    private val templateDir: String
        get() = if (Globals.isKotlin) "templates/kotlin" else "templates"

    fun getResourceTemplate(mainPackage: String, entityName: String): String {
        val mainPackageName = mainPackage.replace("/", ".")
        return getFile("$templateDir/rest.resource.tmpl")
            .replace("{{mainPackageName}}", mainPackageName)
            .replace("{{entityName}}", entityName)
            .replace("{{serviceName}}", entityName.lowercase())
            .replace("{{restServicePrefix}}", "GlobalConstants.CONTEXT_PATH")
    }

    fun getDaoTemplate(mainPackage: String, entityName: String): String {
        val mainPackageName = mainPackage.replace("/", ".")
        val camelRepoName = entityName.substring(0, 1).lowercase() + entityName.substring(1) + "Repository"
        return getFile("$templateDir/dao.tmpl")
            .replace("{{mainPackageName}}", mainPackageName)
            .replace("{{entityName}}", entityName)
            .replace("{{camelRepositoryName}}", camelRepoName)
    }

    fun getFileTemplateByName(mainPackage: String, fileName: String): String {
        val mainPackageName = mainPackage.replace("/", ".")
        return getFile("$templateDir/$fileName.tmpl")
            .replace("{{mainPackageName}}", mainPackageName)
    }

    fun getRepositoryTemplate(mainPackage: String, entityName: String): String {
        val mainPackageName = mainPackage.replace("/", ".")
        return getFile("$templateDir/repository.tmpl")
            .replace("{{mainPackageName}}", mainPackageName)
            .replace("{{entityName}}", entityName)
            .replace("{{primaryKeyType}}", EntityToRESTConstants.PK_DATA_TYPE)
    }

    private fun getFile(fileName: String): String {
        val inputStream = Templates::class.java.classLoader.getResourceAsStream(fileName)!!
        return Scanner(inputStream).use { scanner ->
            buildString {
                while (scanner.hasNextLine()) {
                    appendLine(scanner.nextLine())
                }
            }
        }
    }

    fun getEntityTemplate(entity: Entity, mainPackage: String): String {
        val mainPackageName = mainPackage.replace("/", ".")
        var fileTemplate = getFile("$templateDir/entity.tmpl")
            .replace("{{mainPackageName}}", mainPackageName)
            .replace("{{entityName}}", entity.entityName)
            .replace("{{tableName}}", entity.tableName)

        val sb = StringBuilder()
        if (Globals.isKotlin) {
            buildKotlinEntityFields(sb, entity)
        } else {
            buildJavaEntityFields(sb, entity)
        }
        sb.append("}")

        return fileTemplate + sb.toString()
    }

    private fun buildJavaEntityFields(sb: StringBuilder, entity: Entity) {
        for (column in entity.columns) {
            val indentation = "    "
            if (column.databaseIdType != null) {
                sb.append(indentation).append(column.databaseIdType).append("\n")
            }
            if (column.databaseIdType != null && column.databaseIdType.equals("@ManyToOne", ignoreCase = true)) {
                sb.append(indentation).append("@JoinColumn(name = \"").append(column.databaseColumnName).append("\")\n")
            } else {
                sb.append(indentation).append("@Column(name = \"").append(column.databaseColumnName).append("\")\n")
            }
            sb.append(indentation).append("private ").append(column.javaType).append(" ").append(column.camelCaseFieldName).append(";\n\n")
        }
    }

    private fun buildKotlinEntityFields(sb: StringBuilder, entity: Entity) {
        for (column in entity.columns) {
            val indentation = "    "
            if (column.databaseIdType != null) {
                sb.append(indentation).append(column.databaseIdType).append("\n")
            }
            if (column.databaseIdType != null && column.databaseIdType.equals("@ManyToOne", ignoreCase = true)) {
                sb.append(indentation).append("@JoinColumn(name = \"").append(column.databaseColumnName).append("\")\n")
            } else {
                sb.append(indentation).append("@Column(name = \"").append(column.databaseColumnName).append("\")\n")
            }
            sb.append(indentation).append("var ").append(column.camelCaseFieldName).append(": ").append(column.javaType)
            // Add default values for Kotlin (JPA requires no-arg constructor)
            val default = when (column.javaType) {
                "String" -> " = \"\""
                "Int" -> " = 0"
                "Long" -> " = 0L"
                "Float" -> " = 0f"
                "Double" -> " = 0.0"
                "Boolean" -> " = false"
                "LocalDate" -> "? = null"
                "LocalTime" -> "? = null"
                "LocalDateTime" -> "? = null"
                else -> "? = null" // entity references
            }
            sb.append(default).append("\n\n")
        }
    }
}
